### PR TITLE
General improvement to the C/C++ lexer

### PR DESCRIPTION
**Pull request by hgruniaux**
Improve the Pygments C and C++ lexer. This include adding new features, and correcting olders.

### Changelog
  
  - Add support for hexadecimal floating-point literals (C99 and C++17) (e.g. `0x1.2p3`)
    [Reference](https://en.cppreference.com/w/cpp/language/floating_literal).
  
  - Add support for decimal separators in literals (C++14, but implemented in C lexer too) (e.g. `341'658'362`)
    Decimal separators have been added to the generic C lexer, so C also have this functionality 
    even if it is not in the C specification.
    [Reference](https://en.cppreference.com/w/cpp/language/integer_literal).
    [Reference](https://en.cppreference.com/w/cpp/language/floating_literal).
  
  - Correct integer suffixes. Now only two 'l' (lower or upper cased), and one 'u' (lower or upper cased) 
    are matched. So `1234uuuuuuuu` is no longer valid.
    [Reference](https://en.cppreference.com/w/cpp/language/integer_literal).

  - Add support for `_Alignas`, `_Alignof`, `_Noreturn`, `_Generic`, `_Thread_local`, `_Static_assert`, `_Imaginary`, `_Atomic` (C99, C11).
    [Reference](https://en.cppreference.com/w/c/keyword).

  - Add keywords macro to C lexer (e.g. `complex`, `imaginary`, `noreturn`, `alignas`, etc...)
  
  - Add support for binary integer literals (C++14) (e.g. `0b11001`)
    Binary literals have been added to the generic C lexer, so C also have this functionality 
    even if it is not in the C specification.
    [Reference](https://en.cppreference.com/w/cpp/language/integer_literal).

  - Move `_Bool` and `_Complex` from generic C lexer to C lexer (C++ do not have them).
  
  - Correct floating-point literal suffix, now `l` and `f` both works.
    Previously, even `u` was recognized as a valid suffix.
    [Reference](https://en.cppreference.com/w/cpp/language/floating_literal).
  
  - Correct hexadecimal integer literal prefix, now `0XFF` works.
    Previously, only `0x` (lowercase) worked as a hexadecimal prefix.
    [Reference](https://en.cppreference.com/w/cpp/language/floating_literal).
  
  - Identifiers following `concept` (C++20), `enum`, `struct`, `union`, `class` and `typename`, are now recognized as class names.
    Previously, only identifiers following `class` worked in C++.

  - Add C11 atomic types from standard library highlighting (e.g. `atomic_uint8`).
    Like C99 standard types, C11 atomic types are only highlighted when the `c11highlighting` property is `True` (default).
    [Reference](https://en.cppreference.com/w/c/atomic).

  - Add support for `u'a'`, `U'a'`, and `u8'a'` (before only `L'a'` was supported).
    [Reference](https://en.cppreference.com/w/cpp/language/character_literal) (C++).
    [Reference](https://en.cppreference.com/w/c/language/character_constant) (C).

  - Add support for `$` (dollar sign) in identifiers.
    This is a language extension supported by GCC, Clang, MSVC, and probably many others.

  - Add `_Pragma` keyword to generic C lexer.

  - Add Unicode string literals to C (e.g. `u8"a"`, `u"a"` or `U"a"`).
    Before they were only supported for C++.

  - Add support for Universal Character Names (UCNs)

  - `-` is now highlighted as number in front of numeric literals (e.g. `-9.56` is highlighted as a number, the `-` included).

### Solves

  - #1222
  - #996
  - #906
  - #828 
  - #1162
  - #998 (partial, Unicode characters not yet supported)
  - #1166
  - #1396

### CODE REVIEW

File: pygments/lexers/c_cpp.py [hunk 1]
```diff
@@ -165,10 +165,22 @@ class CFamilyLexer(RegexLexer):
...
+        'atomic_int_least64_t', 'atomic_uint_least64_t', 'atomic_int_fast8_t', 
+        'atomic_uint_fast8_t', 'atomic_int_fast16_t', 'atomic_uint_fast16_t',
+        'atomic_int_fast32_t', 'atomic_uint_fast32_t', 'atomic_int_fast64_t',
+        'atomic_uint_fast64_t', 'atomic_intptr_t', 'atomic_uintptr_t', 'atomic_size_t',
+        'atomic_ptrdiff_t', 'atomic_intmax_t', 'atomic_uintmax_t'}
 
     def __init__(self, **options):
         self.stdlibhighlighting = get_bool_opt(options, 'stdlibhighlighting', True)
         self.c99highlighting = get_bool_opt(options, 'c99highlighting', True)
+        self.c11highlighting = get_bool_opt(options, 'c11highlighting', True)
```
Comment (Anteru) [inline pygments/lexers/c_cpp.py]: This option needs to be documented somewhere.

Comment (HubertGruniaux) [reply pygments/lexers/c_cpp.py]: `stdlibhighlighting`, `c99highlighting` and `platformhighlighting` options don't seem to be documented either. Do you have any idea where the documentation for these options could be added?

Comment (Anteru) [reply pygments/lexers/c_cpp.py]: Uh, that's ... not good :) This should be documented as part of the lexer, like here: https://github.com/pygments/pygments/blob/684975db367f13669aa6d349e083eb8f8458516c/pygments/lexers/email.py#L133

Comment (HubertGruniaux) [reply pygments/lexers/c_cpp.py]: I added the necessary documentation. Not sure it is the right place or the right format, or even if there are mistakes in it.

Comment (Anteru) [reply pygments/lexers/c_cpp.py]: Perfect, thanks a lot for taking care of the other parameters as well! @birkenfeld : Any better idea where to put this?

**Comment by Anteru** [comment]
Thanks a lot for this huge PR -- one question before I assign this to a milestone. There are a few TODO items left, are you planning to work on those as well or should I merge it as-is and leave those open for a future PR?

**Comment by hgruniaux** [comment]
Hi ! I will try to finish the few remaining TODO items, and I will keep you informed when this is over or when I have forsaken.

**Comment by Anteru** [comment]
Perfect. I'm going to cut off 2.6 this week, so I'm targeting this at 2.7. Please take your time to fix this, and thanks a lot for your contribution!

**Comment by hgruniaux** [comment]
This pull request is finished, the remaining TODOs will not be completed, or at least not by me or not in this specific pull-request. So you can merge in the main branch if you want.

Thank you in advance.

**Pull request merged**